### PR TITLE
Headless doc

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This adds headless mode section to the manual and showcases the `:nxref` tag usage in its full power.

There's also documentation for `nyxt/dom` package now.

# Discussion

Some meta-things to discuss while we're at it (not required for merging):
- Should `:nxref` actually get quoted symbols, instead of the unquoted ones it currently gets? This way we can pass variables to it, instead of passing plain class/slot/function/command symbols. More typing, though.
- We should probably be as verbose as this PR is, when we add a new feature. That's why there's a "I have made corresponding changes to the documentation" flag below in the checklist. Because undocumented feature is the feature no one will use.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.  

- [X] I have pulled from master before submitting this PR
- [X] My code follows the style guidelines for Common Lisp code
  - See [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf) and [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml).
- [X] I have performed a self-review of my own code
- [x] My code has been reviewed by at least one peer (the peer review to approve a PR counts.  The reviewer must download and test the code)
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] There are no merge conflicts

### CODE REVIEW

**Comment by aadcg** [comment]
I took the liberty to push some suggestions to this branch. Feel free to revert, squash, change, etc.

Overall, I don't think the manual section is good enough yet. Would you like to read it again and re-work it? I can do it as well, if that doesn't feel offensive.

**Comment by aadcg** [comment]
> We should probably be as verbose as this PR is, when we add a new feature. That's why there's a "I have made corresponding changes to the documentation" flag below in the checklist. Because undocumented feature is the feature no one will use.

Absolutely!

>  Should :nxref actually get quoted symbols, instead of the unquoted ones it currently gets?

I'd lean to say yes, but I trust your judgement here.

File: source/dom.lisp [hunk 1]
```diff
@@ -2,7 +2,25 @@
...
 (nyxt:define-package :nyxt/dom
-  (:documentation "Nyxt-specific DOM classes and functions operating on them."))
+  (:documentation "Nyxt-specific DOM classes and functions operating on them.
+
+The classes are generated for every HTML element, including `h1-element',
+`div-element', `img-element' and others.
+
+Classes corresponding to non-HTML elements, such as `text-element',
+`semantic-element', `h-element' or `input-element', act as higher hierarchical
+entities of element classes.  For instance, every `h<n>-ement' inherits from
```
Comment (Ambrevar) [inline source/dom.lisp:12]: Typo `ement`

File: source/dom.lisp [hunk 2]
```diff
@@ -2,7 +2,25 @@
...
 
 (nyxt:define-package :nyxt/dom
-  (:documentation "Nyxt-specific DOM classes and functions operating on them."))
+  (:documentation "Nyxt-specific DOM classes and functions operating on them.
+
+The classes are generated for every HTML element, including `h1-element',
+`div-element', `img-element' and others.
+
+Classes corresponding to non-HTML elements, such as `text-element',
+`semantic-element', `h-element' or `input-element', act as higher hierarchical
```
Comment (Ambrevar) [inline source/dom.lisp:11]: Isn't` `input` an HTML element?

File: source/manual.lisp [hunk 3]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
 `foo' function:")
     (:pre (:code "nyxt --profile nosave --remote --load foo.lisp --eval '(foo)'"))
 
+    (:h3 "Headless mode")
+    (:p "Much like with scripting functionality, you can use Nyxt without a
```
Comment (Ambrevar) [inline source/manual.lisp:403]: "Much like a scripting..."

File: source/manual.lisp [hunk 4]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+  ((default-new-buffer-url (quri:uri \"https://github.com/atlas-engineer/nyxt\"))))
+
+\(hooks:on *after-startup-hook* ()
+  ;; Once the page's done loading, do your thing.
+  (once-on (buffer-loaded-hook (current-buffer)) (buffer)
+    ;; It's sometimes necessary to sleep, as `buffer-loaded-hook'
+    ;; fires when the page is loaded, which does not mean that all the
+    ;; resources and scripts are done loading yet. Give it some time
+    ;; there.
+    (sleep 0.5)
```
Comment (Ambrevar) [inline source/manual.lisp:437]: No way WebKitGTK can help us here?

File: source/manual.lisp [hunk 5]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+     (:li (:code "--headless") " itself. If you provide it, Nyxt runs
+     headless. If you don't provide " (:code "--headless") ", then Nyxt starts
+     as usual and allows you to see the changes that happen to the page because
+     of your script. A good way to debug and check script behavior, isn't it?")
+     (:li (:code "--no-socket") " flag allows starting as much Nyxt instances as
+     your computer allows you to. Useful to start several instances of Nyxt and
+     do your thing in parallel.")
+     (:li (:code "--profile nosave") " to not pollute your history and cache
+     with the script-accessed pages."))
+
```
Comment (Ambrevar) [inline source/manual.lisp:524]: We _really_ need paging for our manual now! :p

File: source/manual.lisp [hunk 6]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+      as " (:code "hooks:on") ") as a shorthand for the " (:code "nhooks:add-hook") ".")
+      (:li (:nxref :function t nhooks:remove-hook) " (also available
+      as " (:code "hooks:remove-hook") ") that removes the handler from a
+      certain hook.")
+      (:li (:code "hnooks:once-on") " (also available
+      as " (:code "hooks:once-on") ") as a one-shot version
+      of " (:code "nhooks:on") " that removes the handler right after it's
+      completed."))
+     (:li "Operations with the page. For those, we have a whole "
+          (:nxref :package t nyxt/dom) " library and " (:nxref :function t document-model)
```
Comment (Ambrevar) [inline source/manual.lisp:489]: "and the "

File: source/manual.lisp [hunk 7]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+          "buffer slot.")
+     (:ul
+      (:li (:nxref :function t document-model) " method of " (:nxref :class t buffer) " has a
```
Comment (Ambrevar) [inline source/manual.lisp:492]: "The (:nxref ...) method"

File: source/manual.lisp [hunk 8]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+      reasonably fresh copy of the page DOM (Document Object Model, reflects the
+      dynamic structure of the page). It is a Plump DOM, which means that all
```
Comment (Ambrevar) [inline source/manual.lisp:494]: Link to Plump maybe?

File: source/manual.lisp [hunk 9]
```diff
@@ -399,6 +399,129 @@ instance must be non-nil.")
...
+      (:li (:nxref :function t nyxt/dom:select-option-element) " to select an option from the "
+           (:code "<select>") " element options.")))
+    (:p "Additionally, headless mode interacts in useful way with other CLI
+    toggles the Nyxt has:")
+    (:ul
+     (:li (:code "--headless") " itself. If you provide it, Nyxt runs
+     headless. If you don't provide " (:code "--headless") ", then Nyxt starts
+     as usual and allows you to see the changes that happen to the page because
+     of your script. A good way to debug and check script behavior, isn't it?")
+     (:li (:code "--no-socket") " flag allows starting as much Nyxt instances as
```
Comment (Ambrevar) [inline source/manual.lisp:519]: "as many"

**Review by Ambrevar — COMMENTED** [review]
Thanks a lot for updating the manual!  This looks fantastic!

The headless section is quite big though.  I suggest we create a dedicated hook section and move the hook details there.  Then refer to that section from the headless mode section.

File: source/manual.lisp (hunk 4 above)
Comment (aartaka) [reply source/manual.lisp:437]: Nope :(

File: source/manual.lisp (hunk 7 above)
Comment (aartaka) [reply source/manual.lisp:492]: Yep, thought about it too!

**Comment by aartaka** [comment]
@Ambrevar's review addressed. @aadcg, I'm okay with you editing the docs if it ends up being more readable :P

**Comment by aartaka** [comment]
Pagination is beyond the scope of this PR, but I agree it can come in handy. In the meanwhile, we have `help-mode` and `jump-to-heading`.

**Comment by aadcg** [comment]
> Pagination is beyond the scope of this PR, but I agree it can come in handy. 

It's not the right place to start this discussion, but I have serious doubts we actually need it.

**Comment by aadcg** [comment]
@aartaka please avoid commits like 637fe3abc since it makes it harder to squash before merging.

If you agree with all of my latest commits, let me squash this and merge it.

Overall, I still don't like this section, but I don't want to spend more time in it right now.

**Comment by aartaka** [comment]
@aadcg, I'm fine with merging.

Why is https://github.com/atlas-engineer/nyxt/commit/[637fe3a](https://github.com/atlas-engineer/nyxt/commit/637fe3abc0f1368e60ae01312f47148fa38755c2)bc0f1368e60ae01312f47148fa38755c2 that bad?

**Comment by aartaka** [comment]
@aadcg, what exactly do you not like about this section? I'm asking it to avoid in the future docs (that I'm in the mood to write).

**Comment by aadcg** [comment]
On the call with @aartaka I described his manual section as not being "boring enough". We agreed that this material will be published as an article, since it deserves that spotlight. Thanks.

**Comment by Ambrevar** [comment]
Agreed with Artyom.

**Pull request merged**